### Imports

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import urllib
import matplotlib.dates as mdates

### Loading the data in a dataframe

In [32]:
df = pd.read_csv('data.csv')
pd.options.display.max_columns = 999
df

instr  class  nb.repeat  attendance  difficulty  Q1  Q2  Q3  Q4  Q5  Q6  \
0         1      2          1           0           4   3   3   3   3   3   3   
1         1      2          1           1           3   3   3   3   3   3   3   
2         1      2          1           2           4   5   5   5   5   5   5   
3         1      2          1           1           3   3   3   3   3   3   3   
4         1      2          1           0           1   1   1   1   1   1   1   
5         1      2          1           3           3   4   4   4   4   4   4   
6         1      2          1           1           3   4   4   4   4   4   4   
7         1      2          1           1           3   5   5   5   5   5   5   
8         1      2          1           1           3   4   4   4   4   4   4   
9         1      2          1           4           4   4   4   4   4   4   4   
10        1      2          1           3           4   3   3   3   3   3   3   
11        1      2          1           3           3   3   3   3   3   3   3   
12        1      2          1           2           3   4   4   4   4   4   4   
13        1      2          1           1           2   1   1   1   1   1   1   
14        1      2          1           2           3   1   1   1   1   1   1   
15        1      2          1           3           4   5   5   4   4   5   5   
16        1      2          1           3           4   1   1   1   1   1   1   
17        1      2          1           1           2   4   4   4   4   4   4   
18        1      2          1           1           4   1   1   1   1   1   1   
19        1      2          1           2           3   5   5   5   5   5   5   
20        1      2          1           2           4   2   2   2   2   2   2   
21        1      2          1           2           1   1   1   1   1   1   1   
22        1      2          1           3           3   3   3   3   3   3   3   
23        1      2          1           2           4   5   3   3   3   2   2   
24        1      2          1           1           3   5   5   5   5   5   5   
25        1      2          1           0           2   1   1   2   2   2   2   
26        1      2          1           1           2   1   1   1   1   5   5   
27        1      2          1           2           4   2   2   2   2   2   2   
28        1      2          1           3           2   5   5   5   5   5   5   
29        1      2          1           3           3   4   4   4   4   4   4   
...     ...    ...        ...         ...         ...  ..  ..  ..  ..  ..  ..   
5790      3     13          1           4           3   3   3   3   3   3   3   
5791      3     13          1           1           2   4   4   4   4   4   4   
5792      3     13          2           3           3   4   4   4   4   4   4   
5793      3     13          2           4           5   5   5   5   5   5   5   
5794      3     13          2           0           3   2   1   1   1   1   1   
5795      3     13          1           0           1   1   1   1   1   1   1   
5796      3     13          1           1           2   2   2   2   2   2   2   
5797      3     13          1           1           2   3   3   3   3   3   3   
5798      3     13          1           1           4   2   2   2   2   2   2   
5799      3     13          1           0           1   1   1   1   1   1   1   
5800      3     13          2           3           2   4   4   4   4   4   4   
5801      3     13          1           3           3   1   1   1   1   1   1   
5802      3     13          1           0           1   5   5   5   5   5   5   
5803      3     13          1           3           5   2   2   2   1   1   5   
5804      3     13          1           4           3   4   4   4   4   4   4   
5805      3     13          2           1           2   1   1   1   1   1   1   
5806      3     13          2           0           3   4   4   4   4   4   4   
5807      3     13          1        

### Reorganizing the data - Difficulty values
We want to change the 15 scale to -2/2 so that we can have 0 as an absolute value

In [33]:
df['difficulty'] = df['difficulty'].apply(lambda x: x - 3)

### Assessing the class size - Grouping the 13 classes

In [35]:
df.rename(columns={'nb.repeat':'no.taken'}, inplace=True)

In [36]:
# spc means students per class
spc = df.groupby(['class', 'no.taken']).size()
spc

class  no.taken
1      1           272
       2            25
       3             6
2      1           130
       2             7
       3             3
3      1           778
       2            87
       3            39
4      1           154
       2            20
       3            13
5      1           611
       2            28
       3            17
6      1           494
       2            55
       3             9
7      1           150
       2            14
       3            23
8      1           379
       2            69
       3            52
9      1           400
       2            99
       3            72
10     1           425
       2            14
       3             9
11     1           431
       2            43
       3            10
12     1            34
       2             5
       3             2
13     1           651
       2           110
       3            80
dtype: int64

### Renaming the nb.repeats column name to no.taken 
The name is confusing and can be interpreted that a student has taken the calss already once and this
value represents the number of times that he repeated the class

### Assessing how many students have taken the class and how many times

In [37]:
df.groupby('no.taken').size()

no.taken
1    4909
2     576
3     335
dtype: int64

### Checking how many students have taken the class how many times per class

In [38]:
# sapc means student attempts per class
sapc = df.groupby(['class', 'no.taken']).size()
sapc

class  no.taken
1      1           272
       2            25
       3             6
2      1           130
       2             7
       3             3
3      1           778
       2            87
       3            39
4      1           154
       2            20
       3            13
5      1           611
       2            28
       3            17
6      1           494
       2            55
       3             9
7      1           150
       2            14
       3            23
8      1           379
       2            69
       3            52
9      1           400
       2            99
       3            72
10     1           425
       2            14
       3             9
11     1           431
       2            43
       3            10
12     1            34
       2             5
       3             2
13     1           651
       2           110
       3            80
dtype: int64

### Converting the values in percentages to better understand the dataset

In [39]:
# sapc means student attempts per class percentage
sapcp = df.groupby(['class', 'no.taken']).size().reset_index(name='count')
sapcp['Percentage'] = sapcp.groupby('class')['count'].transform(lambda x: x/sum(x) * 100)
sapcp

class  no.taken  count  Percentage
0       1         1    272   89.768977
1       1         2     25    8.250825
2       1         3      6    1.980198
3       2         1    130   92.857143
4       2         2      7    5.000000
5       2         3      3    2.142857
6       3         1    778   86.061947
7       3         2     87    9.623894
8       3         3     39    4.314159
9       4         1    154   82.352941
10      4         2     20   10.695187
11      4         3     13    6.951872
12      5         1    611   93.140244
13      5         2     28    4.268293
14      5         3     17    2.591463
15      6         1    494   88.530466
16      6         2     55    9.856631
17      6         3      9    1.612903
18      7         1    150   80.213904
19      7         2     14    7.486631
20      7         3     23   12.299465
21      8         1    379   75.800000
22      8         2     69   13.800000
23      8         3     52   10.400000
24      9         1    400   70.052539
25      9         2     99   17.338004
26      9         3     72   12.609457
27     10         1    425   94.866071
28     10         2     14    3.125000
29     10         3      9    2.008929
30     11         1    431   89.049587
31     11         2     43    8.884298
32     11         3     10    2.066116
33     12         1     34   82.926829
34     12         2      5   12.195122
35     12         3      2    4.878049
36     13         1    651   77.407848
37     13         2    110   13.079667
38     13         3     80    9.512485

### Getting the mean of the percentages from 1, 2, 3 attemps

In [40]:
sapcp.groupby('no.taken').mean()['Percentage']

no.taken
1    84.848346
2     9.507965
3     5.643689
Name: Percentage, dtype: float64

Based on the previous block, we can assume that the teachers had little to do with students passing/failing.
By calculating the mean for the atempts, we can see that 84.8% of students graduated in the first try, 9.5% in the second try and 5.6% in the third try 

### Split the data in chunks - classes to prepare it for further analysis
This way, we can quickly access the data by class

Data can be accessed like so dfdc[class]

In [41]:
# Create a list of classes
classes = df['class'].unique()
classes
# Create a data frame dictionary to store the entries per class for quick access
# dfd is dataframe dictionary
dfdc = {elem : pd.DataFrame for elem in classes}
for key in dfdc.keys():
    dfdc[key] = df[:][df['class'] == key]


### Split the data in chunks - teachers to prepare it for further analysis
This way, we can quickly access the entries relating to a teacher

Data can be accessed like so dfdt[teacher]

In [30]:
# Create a list of teachers
teachers = df['instr'].unique()
teachers
# Create a data frame dictionary to store the entries per teacher for quick access
# dfdp is dataframe dictionary
dfdp = {elem : pd.DataFrame for elem in teachers}
for key in dfdc.keys():
    dfdp[key] = df[:][df['instr'] == key]

instr  class  no.taken  attendance  difficulty  Q1  Q2  Q3  Q4  Q5  ...  \
0        1      2         1           0           1   3   3   3   3   3  ...   
1        1      2         1           1           0   3   3   3   3   3  ...   
2        1      2         1           2           1   5   5   5   5   5  ...   
3        1      2         1           1           0   3   3   3   3   3  ...   
4        1      2         1           0          -2   1   1   1   1   1  ...   
5        1      2         1           3           0   4   4   4   4   4  ...   
6        1      2         1           1           0   4   4   4   4   4  ...   
7        1      2         1           1           0   5   5   5   5   5  ...   
8        1      2         1           1           0   4   4   4   4   4  ...   
9        1      2         1           4           1   4   4   4   4   4  ...   
10       1      2         1           3           1   3   3   3   3   3  ...   
11       1      2         1           3           0   3   3   3   3   3  ...   
12       1      2         1           2           0   4   4   4   4   4  ...   
13       1      2         1           1          -1   1   1   1   1   1  ...   
14       1      2         1           2           0   1   1   1   1   1  ...   
15       1      2         1           3           1   5   5   4   4   5  ...   
16       1      2         1           3           1   1   1   1   1   1  ...   
17       1      2         1           1          -1   4   4   4   4   4  ...   
18       1      2         1           1           1   1   1   1   1   1  ...   
19       1      2         1           2           0   5   5   5   5   5  ...   
20       1      2         1           2           1   2   2   2   2   2  ...   
21       1      2         1           2          -2   1   1   1   1   1  ...   
22       1      2         1           3           0   3   3   3   3   3  ...   
23       1      2         1           2           1   5   3   3   3   2  ...   
24       1      2         1           1           0   5   5   5   5   5  ...   
25       1      2         1           0          -1   1   1   2   2   2  ...   
26       1      2         1           1          -1   1   1   1   1   5  ...   
27       1      2         1           2           1   2   2   2   2   2  ...   
28       1      2         1           3          -1   5   5   5   5   5  ...   
29       1      2         1           3           0   4   4   4   4   4  ...   
..     ...    ...       ...         ...         ...  ..  ..  ..  ..  ..  ...   
745      1     10         1           3           1   4   4   4   4   4  ...   
746      1     10         1           3           0   3   3   3   3   3  ...   
747      1     10         1           3           0   5   5   5   5   5  ...   
748      1     10         1           4           1   4   4   4   4   4  ...   
749      1     10         1           0          -2   1   1   1   1   1  ...   
750      1     10         1           1          -1   4   4   4   4   4  ...   
751      1     10         1           3           0   2   3   2   3   2  ...   
752      1     10         1           3           0   4   4   4   3   5  ...   
753      1     10         1           1          -1   5   5   5   5   5  ...   
754      1     10         1           0          -2   5   5   5   5   5  ...   
755      1     10         1           4           0   5   5   5   5   5  ...   
756      1     10         1           3           1   3   3   3   3   3  ...   
757      1     10         1           0          -2   5   5   5   5   5  ...   
758      1     10         1           3           0   5   5   5   5   5  ...   
759      1     10         3           4           2   4   4   4   4   4  ...   
760      1     10         1           0          -2   1   1   1   1   1  ...   
761      1     10         1           2           1   4   4   4   4   4  ...   
762      1     10         1           3           0   5   5   5   5   5  ...   
763  

### Important observation

It seems that there are a lot of entries that have the same score in all the question

This might mean that the students have untruthfully graded the questions due to laziness or other factors

Though the factors are not important, we want to assess what is the percentage of entries that have the difference between the entry itself and the mode of the scores of the question 0. At a later date we will determine whether such entries should be treated differently or not.

In [53]:
# First step is to create a dataset that holds only the questions
# qdf represents questions dataframe
qdf = df[df.columns.difference(['instr', 'class', 'no.taken', 'attendance', 'difficulty'])]
qdf

Q1  Q10  Q11  Q12  Q13  Q14  Q15  Q16  Q17  Q18  Q19  Q2  Q20  Q21  Q22  \
0      3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
1      3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
2      5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
3      3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
4      1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5      4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
6      4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
7      5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
8      4    4    5    4    4    4    4    4    4    4    4   4    4    4    4   
9      4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
10     3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
11     3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
12     4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
13     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
14     1    1    1    1    1    1    1    1    2    2    2   1    3    3    3   
15     5    5    5    4    5    5    4    4    5    5    5   5    4    4    5   
16     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
17     4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
18     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
19     5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
20     2    2    2    2    2    2    2    2    2    2    2   2    2    2    2   
21     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
22     3    3    3    3    3    3    3    3    3    3    3   3    2    1    1   
23     5    4    4    5    5    4    3    3    3    4    2   3    2    4    4   
24     5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
25     1    2    2    2    2    2    2    2    2    2    2   1    2    2    2   
26     1    5    5    5    5    5    5    5    5    5    5   1    5    5    5   
27     2    2    2    2    2    2    2    2    2    2    2   2    2    2    2   
28     5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
29     4    4    4    4    4    4    4    4    4    4    4   4    4    4    3   
...   ..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ..  ...  ...  ...   
5790   3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
5791   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5792   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5793   5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
5794   2    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5795   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5796   2    2    2    2    2    2    2    2    3    3    3   2    3    3    3   
5797   3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
5798   2    2    2    2    2    2    2    2    2    2    2   2    2    2    2   
5799   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5800   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5801   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5802   5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
5803   2    4    3    1    3    1    1    1    1    1    1   2    1    1    1   
5804   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5805   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5806   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5807   5    5    5    5    5    5    

In [56]:
# Second step is to calculate the median of the questions per row
qdf['median'] = qdf.median(1)
qdf

/Users/_bogz_/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Q1  Q10  Q11  Q12  Q13  Q14  Q15  Q16  Q17  Q18  Q19  Q2  Q20  Q21  Q22  \
0      3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
1      3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
2      5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
3      3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
4      1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5      4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
6      4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
7      5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
8      4    4    5    4    4    4    4    4    4    4    4   4    4    4    4   
9      4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
10     3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
11     3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
12     4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
13     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
14     1    1    1    1    1    1    1    1    2    2    2   1    3    3    3   
15     5    5    5    4    5    5    4    4    5    5    5   5    4    4    5   
16     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
17     4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
18     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
19     5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
20     2    2    2    2    2    2    2    2    2    2    2   2    2    2    2   
21     1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
22     3    3    3    3    3    3    3    3    3    3    3   3    2    1    1   
23     5    4    4    5    5    4    3    3    3    4    2   3    2    4    4   
24     5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
25     1    2    2    2    2    2    2    2    2    2    2   1    2    2    2   
26     1    5    5    5    5    5    5    5    5    5    5   1    5    5    5   
27     2    2    2    2    2    2    2    2    2    2    2   2    2    2    2   
28     5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
29     4    4    4    4    4    4    4    4    4    4    4   4    4    4    3   
...   ..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ..  ...  ...  ...   
5790   3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
5791   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5792   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5793   5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
5794   2    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5795   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5796   2    2    2    2    2    2    2    2    3    3    3   2    3    3    3   
5797   3    3    3    3    3    3    3    3    3    3    3   3    3    3    3   
5798   2    2    2    2    2    2    2    2    2    2    2   2    2    2    2   
5799   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5800   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5801   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5802   5    5    5    5    5    5    5    5    5    5    5   5    5    5    5   
5803   2    4    3    1    3    1    1    1    1    1    1   2    1    1    1   
5804   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5805   1    1    1    1    1    1    1    1    1    1    1   1    1    1    1   
5806   4    4    4    4    4    4    4    4    4    4    4   4    4    4    4   
5807   5    5    5    5    5    5    

In [57]:
# Third step is to make the difference between each value of the questions and the median
qdf = qdf.apply(lambda x: x - x['median'], axis=1)
qdf

Q1  Q10  Q11  Q12  Q13  Q14  Q15  Q16  Q17  Q18  Q19   Q2  Q20  Q21  \
0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
6     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
7     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
8     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
9     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
10    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
11    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
12    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
13    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
14    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  2.0  2.0   
15    0.0  0.0  0.0 -1.0  0.0  0.0 -1.0 -1.0  0.0  0.0  0.0  0.0 -1.0 -1.0   
16    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
17    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
18    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
19    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
20    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
21    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
22    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 -1.0 -2.0   
23    1.0  0.0  0.0  1.0  1.0  0.0 -1.0 -1.0 -1.0  0.0 -2.0 -1.0 -2.0  0.0   
24    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
25   -1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 -1.0  0.0  0.0   
26   -4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 -4.0  0.0  0.0   
27    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
28    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
29    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
5790  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5791  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5792  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5793  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5794  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5795  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5796  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0   
5797  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5798  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5799  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5800  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5801  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5802  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5803  1.0  3.0  2.0  0.0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
5804  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5805  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5806  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5807  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5808 -1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 -1.0  1.0  0.0   
5809  0.0  0.0  0.0  0.0  0.0

In [73]:
# Fourth step is to count how many of such records where the values of questions after the applied difference is 0
qtc[qtc.apply(pd.value_counts, 1).isna().any(axis=1)]
# df1 = df[df.isna().any(axis=1)]

Q1   Q10   Q11   Q12   Q13   Q14   Q15   Q16   Q17   Q18   Q19    Q2  \
0     True  True  True  True  True  True  True  True  True  True  True  True   
1     True  True  True  True  True  True  True  True  True  True  True  True   
2     True  True  True  True  True  True  True  True  True  True  True  True   
3     True  True  True  True  True  True  True  True  True  True  True  True   
4     True  True  True  True  True  True  True  True  True  True  True  True   
5     True  True  True  True  True  True  True  True  True  True  True  True   
6     True  True  True  True  True  True  True  True  True  True  True  True   
7     True  True  True  True  True  True  True  True  True  True  True  True   
9     True  True  True  True  True  True  True  True  True  True  True  True   
10    True  True  True  True  True  True  True  True  True  True  True  True   
11    True  True  True  True  True  True  True  True  True  True  True  True   
12    True  True  True  True  True  True  True  True  True  True  True  True   
13    True  True  True  True  True  True  True  True  True  True  True  True   
16    True  True  True  True  True  True  True  True  True  True  True  True   
17    True  True  True  True  True  True  True  True  True  True  True  True   
18    True  True  True  True  True  True  True  True  True  True  True  True   
19    True  True  True  True  True  True  True  True  True  True  True  True   
20    True  True  True  True  True  True  True  True  True  True  True  True   
21    True  True  True  True  True  True  True  True  True  True  True  True   
24    True  True  True  True  True  True  True  True  True  True  True  True   
27    True  True  True  True  True  True  True  True  True  True  True  True   
28    True  True  True  True  True  True  True  True  True  True  True  True   
30    True  True  True  True  True  True  True  True  True  True  True  True   
31    True  True  True  True  True  True  True  True  True  True  True  True   
33    True  True  True  True  True  True  True  True  True  True  True  True   
36    True  True  True  True  True  True  True  True  True  True  True  True   
37    True  True  True  True  True  True  True  True  True  True  True  True   
38    True  True  True  True  True  True  True  True  True  True  True  True   
39    True  True  True  True  True  True  True  True  True  True  True  True   
41    True  True  True  True  True  True  True  True  True  True  True  True   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
5781  True  True  True  True  True  True  True  True  True  True  True  True   
5783  True  True  True  True  True  True  True  True  True  True  True  True   
5784  True  True  True  True  True  True  True  True  True  True  True  True   
5785  True  True  True  True  True  True  True  True  True  True  True  True   
5787  True  True  True  True  True  True  True  True  True  True  True  True   
5788  True  True  True  True  True  True  True  True  True  True  True  True   
5789  True  True  True  True  True  True  True  True  True  True  True  True   
5790  True  True  True  True  True  True  True  True  True  True  True  True   
5791  True  True  True  True  True  True  True  True  True  True  True  True   
5792  True  True  True  True  True  True  True  True  True  True  True  True   
5793  True  True  True  True  True  True  True  True  True  True  True  True   
5795  True  True  True  True  True  True  True  True  True  True  True  True   
5797  True  True  True  True  True  True  True  True  True  True  True  True   
5798  True  True  True  True  True  True  True  True  True  True  True  True   
5799  True  True  True  True  True  True  True  True  True  True  True  True   
5800  True  True  True  True  True  True  True  True  True  True  True  True   
5801  True  True  True  True  True  True  True  True  True  True  True  True   
5802  True  True  True  True  True  True  True  True  True  True  True  True   
5804  Tr

In [75]:
100 *2985/5819 

51.297473792747894

In [ ]:
## 51% of the values are all equal. What do we do now...